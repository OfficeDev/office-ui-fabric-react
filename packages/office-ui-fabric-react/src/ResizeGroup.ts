export * from './packages/react-commanding/ResizeGroup';
