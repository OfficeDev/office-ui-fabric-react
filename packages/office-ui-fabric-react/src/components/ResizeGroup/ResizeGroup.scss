@import '../../common/common';

.root {
  display: block;
}

.measured {
  flex-direction: horizontal;
  flex-wrap: nowrap;
  display: flex;
  position: absolute;
  visibility: hidden;
}