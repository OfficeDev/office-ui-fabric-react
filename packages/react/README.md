# @uifabric/react

**React components for [Office UI Fabric React](https://dev.microsoft.com/fabric)**

These are not production-ready components and **should never be used in product**. This space is useful for testing new components whose APIs might change before final release.

To import React components:

```js
import { ComponentName } from '@uifabric/react';
```

Once the React component graduates to a production release, the component will be available at:

```js
import { ComponentName } from 'office-ui-fabric-react';
```
