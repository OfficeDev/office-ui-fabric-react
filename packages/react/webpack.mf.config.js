const path = require('path');
const fs = require('fs');
const webpack = require('webpack');
const getResolveAlias = require('@fluentui/scripts/webpack/getResolveAlias');
const { createServeConfig } = require('@fluentui/scripts/webpack/webpack-resources');
const { webpackMerge } = require('just-scripts');

const shared = {
  react: { singleton: true },
  'react-dom': { singleton: true },
};

const rootComponents = fs
  .readdirSync(path.join(__dirname, 'src'))
  .filter(filename => filename[0].toUpperCase() === filename[0])
  .map(filename => filename.replace(/\.tsx?/, ''));
const rootComponentsExposes = {};
for (const component of rootComponents) {
  rootComponentsExposes[`./lib/${component}`] = `@fluentui/react/src/${component}`;
}

// TODO: fork-ts-checker-webpack-plugin needs to be taught about the resolve alias -> pathmapping in order for the type checking to work
process.env.SKIP_TYPECHECK = true;

const myConfig = {
  output: {},
  entry: './empty-entrypoint.js',
  mode: 'development',
  devtool: 'cheap-module-source-map',
  optimization: {
    concatenateModules: false,
  },
  module: {
    rules: [],
  },
  plugins: [
    new webpack.container.ModuleFederationPlugin({
      name: 'fluentuiReact',
      filename: 'remoteEntry.js',
      exposes: {
        ...rootComponentsExposes,
        './lib/compat/Button': '@fluentui/react/src/compat/Button',
        '.': '@fluentui/react/src/index',
        './lib/index': '@fluentui/react/src/index',
      },
      shared,
    }),
  ],
  resolve: {
    alias: getResolveAlias(),
  },
  devServer: {
    port: 2345,
    static: './dist',
  },
};

const metaConfig = webpackMerge.merge(createServeConfig(), myConfig);

module.exports = metaConfig;
