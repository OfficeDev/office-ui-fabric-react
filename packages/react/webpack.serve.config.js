const path = require('path');
const resources = require('@uifabric/build/webpack/webpack-resources');

module.exports = resources.createServeConfig({
  entry: './src/demo/index.tsx',

  output: {
    filename: 'demo-app.js',
  },

  externals: {
    react: 'React',
    'react-dom': 'ReactDOM',
  },

  resolve: {
    alias: {
      '@fluentui/react/src': path.join(__dirname, 'src'),
      '@fluentui/react/lib': path.join(__dirname, 'lib'),
      '@fluentui/react': path.join(__dirname, 'lib'),
      'Props.ts.js': 'Props',
      'Example.tsx.js': 'Example',
    },
  },
});
