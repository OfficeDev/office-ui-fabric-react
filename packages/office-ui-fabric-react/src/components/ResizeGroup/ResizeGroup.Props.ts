import * as React from 'react';
import { ResizeGroup } from './ResizeGroup';


export interface IResizeGroup {
  measure?: () => void;
}

export interface IResizeGroupProps extends React.HTMLProps<ResizeGroup | HTMLElement> {
  data?: any;
  onRenderData?: (data: any) => JSX.Element;
  onReduceData?: (prevData: any) => any;
  componentRef?: (component: IResizeGroup) => void;
}
