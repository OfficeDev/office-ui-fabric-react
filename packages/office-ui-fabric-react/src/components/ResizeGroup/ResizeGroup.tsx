import * as React from 'react';
import {
  css,
  BaseComponent
} from '../../Utilities';
import { IResizeGroupProps } from './ResizeGroup.Props';
import styles = require('./ResizeGroup.scss');

const RESIZE_DELAY = 16;

export interface IResizeGroupState {

  /**
   * Final data used to render proper sized component
  */
  renderedData?: any;

  /**
   * A flag to determine if a new measurement should be made upon state change
  */
  shouldMeasure?: boolean;

  /**
   * Data to render in a hidden div for measurement
   */
  dataToMeasure?: any;

  /**
   * Are we resizing to accomodate having more or less available space?
   * The 'grow' direction is when the container may have more room than the last render,
   * such as when a window resize occurs. This means we will try to fit more content in the window.
   * The 'shrink' direction is when the contents don't fit in the container and we need
   * to find a transformation of the data that makes everything fit.
   */
  resizeDirection?: 'grow' | 'shrink';

  /**
   * Set to true when the content container might have new dimensions and should
   * be remeasured.
   */
  measureContainer?: boolean;
}

const getCachedMeasurementProvider = () => {
  const measurementsCache: { [key: string]: number } = {};

  return {
    getCachedMeasurement: (data: any): number | undefined => {
      if (data && data.cacheKey && measurementsCache.hasOwnProperty(data.cacheKey)) {
        return measurementsCache[data.cacheKey];
      }

      return undefined;
    },
    addMeasurementToCache: (data: any, measurement: number): void => {
      if (data.cacheKey) {
        measurementsCache[data.cacheKey] = measurement;
      }
    }
  };
};

const getCachedContentMeasurer = () => {
  const _cachedMeasurementProvider = getCachedMeasurementProvider();
  let _containerWidth: number | undefined;

  const _getMeasuredWidth = (measuredData: any, elementToMeasure: HTMLElement | null): number => {
    let cachedWidth = _cachedMeasurementProvider.getCachedMeasurement(measuredData);
    if (cachedWidth !== undefined) {
      return cachedWidth;
    }

    let measuredWidth = elementToMeasure.getBoundingClientRect().width;
    _cachedMeasurementProvider.addMeasurementToCache(measuredData, measuredWidth);
    return measuredWidth;
  };

  const _shrinkContentsUntilTheyFit = (data: any,
    onReduceData: (prevData: any) => any,
    elementToMeasure: HTMLElement | null): IResizeGroupState => {

    let dataToMeasure = data;
    let measuredWidth = _getMeasuredWidth(data, elementToMeasure);
    while (measuredWidth > _containerWidth) {
      let nextMeasuredData = onReduceData(dataToMeasure);

      // We don't want to get stuck in an infinite render loop when there are no more
      // scaling steps, so implementations of onReduceData should return undefined when
      // there are no more scaling states to apply.
      if (nextMeasuredData === undefined) {
        return {
          renderedData: dataToMeasure
        };
      }

      let cachedWidth = _cachedMeasurementProvider.getCachedMeasurement(nextMeasuredData);

      // If the measurement isn't in the cache, we need to rerender with some data in a hidden div
      if (cachedWidth === undefined) {
        return {
          dataToMeasure: nextMeasuredData
        };
      }

      measuredWidth = cachedWidth;
      dataToMeasure = nextMeasuredData;
    }

    return {
      renderedData: dataToMeasure,
      dataToMeasure: undefined
    };
  };

  const _growDataUntilItDoesNotFit = (data: any,
    onGrowData: (prevData: any) => any,
    elementToMeasure: HTMLElement | null): IResizeGroupState => {

    let dataToMeasure = data;
    let measuredWidth = _getMeasuredWidth(data, elementToMeasure);

    while (measuredWidth < _containerWidth) {
      let nextMeasuredData = onGrowData(dataToMeasure);

      // We don't want to get stuck in an infinite render loop when there are no more
      // scaling steps, so implementations of onReduceData should return undefined when
      // there are no more scaling states to apply.
      if (nextMeasuredData === undefined) {
        return {
          renderedData: dataToMeasure
        };
      }

      let cachedWidth = _cachedMeasurementProvider.getCachedMeasurement(nextMeasuredData);

      // If the measurement isn't in the cache, we need to rerender with some data in a hidden div
      if (cachedWidth === undefined) {
        return {
          dataToMeasure: nextMeasuredData
        };
      }

      measuredWidth = cachedWidth;
      dataToMeasure = nextMeasuredData;
    }

    return {
      dataToMeasure,
      resizeDirection: 'shrink'
    };
  };

  return {
    getNextResizeGroupState: (data: any,
      onReduceData: (prevData: any) => any,
      onGrowData: (prevData: any) => any | undefined,
      elementToMeasure: HTMLElement | null,
      direction: 'grow' | 'shrink'): IResizeGroupState => {

      if (direction === 'grow') {
        return _growDataUntilItDoesNotFit(data, onGrowData, elementToMeasure);
      } else {
        return _shrinkContentsUntilTheyFit(data, onReduceData, elementToMeasure);
      }
    },
    updateContainerWidth: (newWidth: number, fullWidthData: any, renderedData: any, hasOnGrowData: boolean): IResizeGroupState => {
      let nextState: IResizeGroupState;
      if (_containerWidth && newWidth > _containerWidth) {
        if (hasOnGrowData && renderedData) {
          nextState = {
            resizeDirection: 'grow',
            dataToMeasure: renderedData
          };
        } else {
          nextState = {
            resizeDirection: 'shrink',
            dataToMeasure: fullWidthData
          };
        }
      } else {
        nextState = {
          resizeDirection: 'shrink',
          dataToMeasure: renderedData || fullWidthData
        };
      }
      _containerWidth = newWidth;
      return { ...nextState, measureContainer: false };
    }
  };
};

export class ResizeGroup extends BaseComponent<IResizeGroupProps, IResizeGroupState> {
  private _measurementProvider = getCachedContentMeasurer();
  private _root: HTMLElement;
  private _measured: HTMLElement;

  constructor(props: IResizeGroupProps) {
    super(props);
    this.state = {
      dataToMeasure: { ...this.props.data },
      resizeDirection: 'shrink',
      measureContainer: true,
      renderedData: null
    };
  }

  public render() {
    const { onRenderData, data } = this.props;
    const { dataToMeasure, renderedData } = this.state;

    if (Object.keys(data).length === 0) {
      return null;
    }

    return (
      <div className={ css('ms-ResizeGroup') } ref={ this._resolveRef('_root') }>
        { dataToMeasure && (
          <div className={ css(styles.measured) } ref={ this._resolveRef('_measured') }>
            { onRenderData(dataToMeasure) }
          </div>
        ) }

        { renderedData && onRenderData(renderedData) }
      </div>
    );
  }

  public componentDidMount() {
    this._afterComponentRendered();
    this._events.on(window, 'resize', this._async.debounce(this._onResize, RESIZE_DELAY, { leading: true }));
  }

  public componentWillReceiveProps(nextProps: IResizeGroupProps) {
    this.setState({
      resizeDirection: 'shrink',
      dataToMeasure: { ...nextProps.data },
      measureContainer: true // Receiving new props means the parent might rerender and the root width might change
    });
  }

  public componentDidUpdate(prevProps: IResizeGroupProps) {
    if (this.state.renderedData) {
      if (this.props.dataDidRender) {
        this.props.dataDidRender(this.state.renderedData);
      }
    }
    this._afterComponentRendered();
  }

  private _afterComponentRendered() {
    if (this.state.measureContainer) {
      this.setState(
        this._measurementProvider.updateContainerWidth(this._root.getBoundingClientRect().width,
          this.props.data,
          this.state.renderedData,
          !!this.props.onGrowData));
    } else if (this.state.dataToMeasure) {
      this.setState(this._measurementProvider.getNextResizeGroupState(this.state.dataToMeasure,
        this.props.onReduceData,
        this.props.onGrowData,
        this._measured,
        this.state.resizeDirection));
    }
  }

  private _onResize() {
    if (this._root) {
      this.setState({ measureContainer: true });
    }
  }
}