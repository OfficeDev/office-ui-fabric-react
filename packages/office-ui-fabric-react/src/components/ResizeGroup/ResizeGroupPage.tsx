import * as React from 'react';
import { Link } from 'office-ui-fabric-react/lib/Link';
import { LayerHost } from 'office-ui-fabric-react/lib/Layer';
import {
  ExampleCard,
  ComponentPage,
  PropertiesTableSet
} from '@uifabric/example-app-base';
import { ResizeGroupOverflowSetExample } from './examples/ResizeGroup.OverflowSet.Example';

const ResizeGroupBasicExampleCode = require('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/examples/ResizeGroup.OverflowSet.Example.tsx') as string;

export class ResizeGroupPage extends React.Component<any, any> {
  public render() {
    return (
      <ComponentPage
        title='ResizeGroup'
        componentName='ResizeGroupExample'
        exampleCards={
          <LayerHost>
            <ExampleCard title='ResizeGroup' code={ ResizeGroupBasicExampleCode }>
              <ResizeGroupOverflowSetExample />
            </ExampleCard>
          </LayerHost>
        }
        propertiesTables={
          <PropertiesTableSet
            sources={ [
              require<string>('!raw-loader!office-ui-fabric-react/src/components/ResizeGroup/ResizeGroup.Props.ts')
            ] }
          />
        }
        overview={
          <div>
            <Link target='_blank' href='http://dev.office.com/fabric/components/ResizeGroup'>ResizeGroup</Link>
            <span> supplement content associated with a specific UI component.</span>
          </div>
        }
        dos={
          <div>
            <ul>
              <li>
                Ensure the width of the parent of this component has a fixed width that does not depend on the dimensions of it's children.
                Failure to do so may result in ResizeGroup attempting to fill a width of 0px.
              </li>
              <li>Include a cacheKey in your data to improve performance </li>
            </ul>
          </div>
        }
        donts={
          <div>
            <ul>
              <li>Do any DOM measurements inside your onReduce function as this will degrade performance</li>
              <li>Provide too many different return values for onReduce, it will degrade performance</li>
            </ul>
          </div>
        }
      />
    );
  }
}
