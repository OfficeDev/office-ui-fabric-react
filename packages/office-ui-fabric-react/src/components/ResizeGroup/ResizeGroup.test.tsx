import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { expect } from 'chai';
import { ResizeGroup, IResizeGroupState, getNextResizeGroupStateProvider, getMeasurementCache } from './ResizeGroup';
import { IResizeGroupProps } from './ResizeGroup.Props';
import * as sinon from 'sinon';

interface ITestScalingData {
  scalingIndex: number;
}

function onReduceScalingData(data: ITestScalingData): ITestScalingData {
  return {
    scalingIndex: data.scalingIndex - 1
  };
}

function getRequiredResizeGroupProps() {
  return {
    data: {},
    onReduceData: sinon.stub(),
    onRenderData: sinon.stub()
  };
}

describe('ResizeGroup', () => {
  it('renders the result of onRenderData', () => {
    const initialData = { content: 5 };
    const renderedDataId = 'onRenderDataId';
    const onRenderData = (data: any) => <div id={ renderedDataId }> Rendered data: { data.content }</div >;

    const wrapper = shallow<IResizeGroupProps, IResizeGroupState>(
      <ResizeGroup
        data={ initialData }
        onReduceData={ onReduceScalingData }
        onRenderData={ onRenderData }
      />
    );

    expect(wrapper.containsMatchingElement(onRenderData(initialData))).to.equal(true);
  });

  describe('getNextResizeGroupStateProvider', () => {
    it('does not provide a new state when there is no container width provided or data to measure', () => {
      const resizeGroupProps = getRequiredResizeGroupProps();
      const resizeGroupState: IResizeGroupState = {};
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        getMeasuredElementWidthStub);

      expect(result).to.equal(undefined);
      expect(getMeasuredElementWidthStub.callCount).to.equal(0);
    });

    it('sets the renderedData when the contents fit', () => {
      const dataToMeasure = { foo: 'bar' };
      const resizeGroupProps = getRequiredResizeGroupProps();
      const resizeGroupState: IResizeGroupState = { dataToMeasure, resizeDirection: 'shrink' };
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();
      getMeasuredElementWidthStub.returns(25);

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        getMeasuredElementWidthStub,
        50);

      expect(result).to.deep.equal({
        renderedData: dataToMeasure,
        measureContainer: false,
        dataToMeasure: undefined,
        resizeDirection: undefined
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(1);
    });

    it('calls onReduceData and sets the next measuredData when contents do not fit', () => {
      const dataToMeasure = { index: 5 };
      const resizeGroupProps = getRequiredResizeGroupProps();
      const resizeGroupState: IResizeGroupState = { dataToMeasure, resizeDirection: 'shrink' };
      resizeGroupProps.onReduceData.returns({ index: 4 });
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();
      getMeasuredElementWidthStub.returns(25);

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        getMeasuredElementWidthStub,
        10);

      expect(result).to.deep.equal({
        measureContainer: false,
        dataToMeasure: { index: 4 },
        resizeDirection: 'shrink'
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(1);
    });

    it('does not call getMeasuredElementBounds when the data has already been cached', () => {
      const dataToMeasure = { index: 5, cacheKey: 'foo' };

      let measurementCache = getMeasurementCache();
      measurementCache.addMeasurementToCache(dataToMeasure, 40);
      const getNextResizeGroupState = getNextResizeGroupStateProvider(measurementCache);

      const resizeGroupProps = getRequiredResizeGroupProps();
      const resizeGroupState: IResizeGroupState = { dataToMeasure, resizeDirection: 'shrink' };
      const measuredElementWidthStub = sinon.stub();

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        measuredElementWidthStub,
        100);

      expect(result).to.deep.equal({
        renderedData: dataToMeasure,
        measureContainer: false,
        dataToMeasure: undefined,
        resizeDirection: undefined
      });
      expect(measuredElementWidthStub.callCount).to.equal(0);
    });

    it('calls onReduceData multiple times when everything is in the cache', () => {
      const dataArray = [{ cacheKey: '5' },
      { cacheKey: '4' },
      { cacheKey: '3' }];

      let measurementCache = getMeasurementCache();
      measurementCache.addMeasurementToCache(dataArray[0], 50);
      measurementCache.addMeasurementToCache(dataArray[1], 40);
      measurementCache.addMeasurementToCache(dataArray[2], 5);
      const getNextResizeGroupState = getNextResizeGroupStateProvider(measurementCache);

      const resizeGroupProps = getRequiredResizeGroupProps();
      resizeGroupProps.onReduceData.onFirstCall().returns(dataArray[1]);
      resizeGroupProps.onReduceData.onSecondCall().returns(dataArray[2]);

      const resizeGroupState: IResizeGroupState = { dataToMeasure: dataArray[0], resizeDirection: 'shrink' };
      const measuredElementWidthStub = sinon.stub();

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        measuredElementWidthStub,
        10);

      expect(result).to.deep.equal({
        renderedData: dataArray[2],
        measureContainer: false,
        dataToMeasure: undefined,
        resizeDirection: undefined
      });
      expect(measuredElementWidthStub.callCount).to.equal(0);
    });

    it('sets dataToMeasure when the current data is in the cache but the onReduceData result is not in the cache', () => {
      const dataArray = [{ cacheKey: '5' },
      { cacheKey: '4' }];

      let measurementCache = getMeasurementCache();
      measurementCache.addMeasurementToCache(dataArray[0], 50);
      const getNextResizeGroupState = getNextResizeGroupStateProvider(measurementCache);

      const resizeGroupProps = getRequiredResizeGroupProps();
      resizeGroupProps.onReduceData.onFirstCall().returns(dataArray[1]);

      const resizeGroupState: IResizeGroupState = { dataToMeasure: dataArray[0], resizeDirection: 'shrink' };
      const measuredElementWidthStub = sinon.stub();

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        measuredElementWidthStub,
        10);

      expect(result).to.deep.equal({
        dataToMeasure: dataArray[1],
        measureContainer: false,
        resizeDirection: 'shrink'
      });
      expect(measuredElementWidthStub.callCount).to.equal(0);
    });

    it('renders the last measured data if onReduceData returns undefined', () => {
      const dataToMeasure = { index: 5 };
      const resizeGroupProps = getRequiredResizeGroupProps();

      const resizeGroupState: IResizeGroupState = { dataToMeasure, resizeDirection: 'shrink' };
      resizeGroupProps.onReduceData.returns(undefined);

      const getNextResizeGroupState = getNextResizeGroupStateProvider();

      const getMeasuredElementWidthStub = sinon.stub();
      getMeasuredElementWidthStub.returns(25);

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        getMeasuredElementWidthStub,
        10);

      expect(result).to.deep.equal({
        dataToMeasure: undefined,
        renderedData: dataToMeasure,
        measureContainer: false,
        resizeDirection: undefined
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(1);
    });

    it('renders the last measured data in the cache if onReduceData returns undefined', () => {
      const dataArray = [{ cacheKey: '5' },
      { cacheKey: '4' }];

      let measurementCache = getMeasurementCache();
      measurementCache.addMeasurementToCache(dataArray[0], 50);
      measurementCache.addMeasurementToCache(dataArray[1], 40);
      const getNextResizeGroupState = getNextResizeGroupStateProvider(measurementCache);

      const resizeGroupProps = getRequiredResizeGroupProps();
      resizeGroupProps.onReduceData.onFirstCall().returns(dataArray[1]);
      resizeGroupProps.onReduceData.onSecondCall().returns(undefined);

      const resizeGroupState: IResizeGroupState = { dataToMeasure: dataArray[0], resizeDirection: 'shrink' };
      const measuredElementWidthStub = sinon.stub();

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        measuredElementWidthStub,
        10);

      expect(result).to.deep.equal({
        dataToMeasure: undefined,
        renderedData: dataArray[1],
        measureContainer: false,
        resizeDirection: undefined,
      });
      expect(measuredElementWidthStub.callCount).to.equal(0);
    });

    it('does not crash when the container size is set and there is no dataToMeasure', () => {
      const dataToMeasure = { foo: 'bar' };
      const resizeGroupProps = getRequiredResizeGroupProps();
      const resizeGroupState: IResizeGroupState = { dataToMeasure, resizeDirection: 'shrink', };
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();

      let result = getNextResizeGroupState(resizeGroupProps,
        {},
        getMeasuredElementWidthStub,
        50);

      expect(result).to.deep.equal({
        measureContainer: false,
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(0);
    });

    it('makes sure the contents still fit when the container width decreases', () => {
      const initialWidth = 50;
      const reducedWidth = 40;
      const renderedData = { foo: 'bar' };
      const resizeGroupProps = getRequiredResizeGroupProps();
      const resizeGroupState: IResizeGroupState = { renderedData, resizeDirection: 'shrink', };
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();

      // Set the initial window width
      getNextResizeGroupState(resizeGroupProps,
        {},
        undefined,
        initialWidth);

      // Pass in a state that reflects some rendered data
      let currentState = {
        renderedData: renderedData
      };

      let result = getNextResizeGroupState(resizeGroupProps,
        currentState,
        getMeasuredElementWidthStub,
        reducedWidth);

      // Important to note that we do not start scaling from the initial data,
      // we continue from the last rendered data.
      expect(result).to.deep.equal({
        renderedData: renderedData,
        dataToMeasure: renderedData,
        measureContainer: false,
        resizeDirection: 'shrink'
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(0);
    });

    it('starts from the beginning when the container width increases', () => {
      const initialWidth = 50;
      const increasedWidth = 60;
      const renderedData = { foo: 'bar' };
      const resizeGroupProps = { ...getRequiredResizeGroupProps(), data: { foo: 'initialData' } };
      const resizeGroupState: IResizeGroupState = { renderedData, resizeDirection: 'shrink' };
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();

      // Set the initial window width
      getNextResizeGroupState(resizeGroupProps,
        {},
        undefined,
        initialWidth);

      // Pass in a state that reflects some rendered data
      let currentState = {
        renderedData: renderedData
      };

      let result = getNextResizeGroupState(resizeGroupProps,
        currentState,
        getMeasuredElementWidthStub,
        increasedWidth);

      expect(result).to.deep.equal({
        renderedData: renderedData,
        dataToMeasure: resizeGroupProps.data,
        resizeDirection: 'shrink',
        measureContainer: false
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(0);
    });

    it('renders contents when the resizeDirection is grow, there is no onGrowData, and the contents fit', () => {
      const dataToMeasure = { foo: 'bar' };
      const resizeGroupProps = getRequiredResizeGroupProps();
      const resizeGroupState: IResizeGroupState = { dataToMeasure, resizeDirection: 'grow' };
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();
      getMeasuredElementWidthStub.returns(25);

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        getMeasuredElementWidthStub,
        50);

      expect(result).to.deep.equal({
        renderedData: dataToMeasure,
        measureContainer: false,
        dataToMeasure: undefined,
        resizeDirection: undefined
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(1);
    });

    it('measures the next state when the resizeDirection is grow and the dataToMeasure fits', () => {
      const dataToMeasure = { index: 1 };
      const onGrowData = sinon.stub();
      onGrowData.returns({ index: 2 });
      const resizeGroupProps = { ...getRequiredResizeGroupProps(), onGrowData };
      const resizeGroupState: IResizeGroupState = { dataToMeasure, resizeDirection: 'grow' };
      const getNextResizeGroupState = getNextResizeGroupStateProvider();
      const getMeasuredElementWidthStub = sinon.stub();
      getMeasuredElementWidthStub.returns(25);

      let result = getNextResizeGroupState(resizeGroupProps,
        resizeGroupState,
        getMeasuredElementWidthStub,
        50);

      expect(result).to.deep.equal({
        measureContainer: false,
        dataToMeasure: { index: 2 },
        resizeDirection: 'grow'
      });
      expect(getMeasuredElementWidthStub.callCount).to.equal(1);
      expect(onGrowData.callCount).to.equal(1);
    });
  });
});
