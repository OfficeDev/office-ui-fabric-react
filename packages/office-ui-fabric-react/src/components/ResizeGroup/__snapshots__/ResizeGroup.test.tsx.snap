// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizeGroup renders the ResizeGroup correctly 1`] = `
<div
  className=
      ms-ResizeGroup
      TestClassName
      {
        display: block;
        position: relative;
      }
>
  <div
    className=
        ms-ResizeGroup--measured
        {
          position: fixed;
          visibility: hidden;
        }
  >
    <div
      id="onRenderDataId"
    >
       Rendered data: 
      5
    </div>
  </div>
</div>
`;
